### DISCUSSION TITLE
How to recognise a keypress in Ctrlr

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#256] [Q&A]
Part of the panel under construction has a numeric pad, see picture. All the "keys" are at the moment operated with a mousedown message. Is it possible to mimic this with actual keypresses on the keyboard?

<img width="222" alt="Screenshot 2021-04-14 at 20 42 30" src="https://user-images.githubusercontent.com/11859363/114763283-eb999980-9d62-11eb-816b-bacc22cf46be.png">

**Comment by DRBAUS** [comment]
By keyboard you mean the computer keyboard, I assume?!

Dat zou moeten kunnen denk ik! 
(That should be possible I think!)

When you use a uiCustomComponent then under Component you will find a 'Key down callback'. This might be what you're looking for.
I just looked it up because you asked. I have not used this or tried this myself...yet.

**Reply by bijlevel** [reply]
Thanks/Dank u! (You're right I should have mentioned it was the computer keyboard. In the future I will distinguish between a computer and instrument keyboard :-).) Question: how do you look up such "behaviour" of a component?

**Reply by DRBAUS** [reply]
I had not seen it before in any other component so I checked with the uiCustomComponent if it allows for something like this.

**Comment by Ted-juh** [comment]
There is a demo panel about keyboard events. Conveniently named: DEMO-Keyboard events.panel.  It uses a customComponent where you can add a function to the keyDown and keyStateChanged callbacks. You could use it as a transparent overlay. But the one downside I see is that in the demo-panel, the customcomponent needs to be clicked "once" to get focus for the keyPress events. If you can find a solution to that then you got your actual keypresses.

**Reply by bijlevel** [reply]
Thanks! I will dive into it and who knows I manage to achieve what I want :-).

**Reply by bijlevel** [reply]
@DnaldeT I understand the demo and in my case I need indeed a mouse click on the uiCustomComponent. So the inevitable question: can one simulate a mouse click on a component? Or is there another way to set the focus on a component?

**Comment by DRBAUS** [comment]
I have looked at the Demo Panel.

I noticed that both the "mouse down grabs keyboard focus" and "this component need keyboard focus" are checked.

I unchecked the "mouse down grabs keyboard focus" and nothing happens....until I pressed the TAB key.

Pressing TAB once seems to put the focus on the component and I am able to type individual letters. Then pressing TAB takes the focus from the component and nothing happens. Then pressing TAB 2 times puts the focus back on the component and I can type again.

This may not be of any help but it was an observation and I thought I would share it because it contains both the yes and no answer. 

edit: actually you can do this with the "mouse down grabs keyboard focus" checked as well.

**Reply by bijlevel** [reply]
Thanks for the input. I couldn't reproduce this TAB behaviour consistently. Sometimes is seems to work, but most of the time it just registered (shows) the Tab keypress. This is with the demo panel, but with an another button added to simulate my own panel.

**Reply by Ted-juh** [reply]
Looks like the tab button is to switch between (child?) components. You can test this when you add a couple of sliders. If you press tab, you can see that it selects the label of a Slider or label (if click to edit is enabled). Tap Tab again and it jumps to the next slider or label in order.

**Comment by Ted-juh** [comment]
Is there a label involved to see what's being typed when you use the numeric pad with the mouse? If that's the case, you can use the label itself to type directly into it with the keyboard after clicking it (or double click). 

Simulate a mouse click, I guess that's a no. It is possible to set focus on a component. There are a few .def's in the LComponents.cpp. (setWantsKeyboardFocus, getWantsKeyBoardFocus, grabKeyBoardFocus, etc.).  And it might be possible to do it with a keyPresEvent. Components have a function "addKeyListener". The component is listening for a certain keyPress then. Maybe you can use keyPressEvent in combination with one of the .defs to set the focus to the numeric pad when pressing a certain keyPress. 

Maybe the Global Keyboard Events demo panel might help as well. It uses the keyPressEvent and originatingComponent. Or take a look at "focusTraverser" Sorry to be not that much of a help. This is very new territory for me as well. I think it's less hassle to just make someone click the numeric pad with the mouse, it sets the focus to the numeric pad and will catch what's being typed.

**Reply by bijlevel** [reply]
I will look into the various options you mention and come back to this. Bedankt!

**Reply by bijlevel** [reply]
I didn't make much progress, but maybe I should make clear what I'm after. I have this kind of background on my panel (showing just a portion):

<img width="369" alt="Screenshot 2021-06-03 at 17 42 23" src="https://user-images.githubusercontent.com/11859363/120673032-1e265f80-c493-11eb-998b-d9750a4a3e49.png">

All the numbers (and also the "+', "-" signs and the arrows on the left) are covered by transparant buttons. When the users clicks on a button, some (midi) action is performed. So far nothing special ;-). What I want to achieve is when the user presses a (number) key or for example the arrow keys (or even some other key) on the computer keyboard, I want the panel to perform the same action as when the user has pressed the corresponding button. BUT both ways should be possible: using the mouse AND using the computer keyboard alternating. And that seems to be the problem! I can provide a switch so it is either the computer keyboard as input or the mouse, but never both ways at the same time. And it is not really user friendly, because as soon as one clicks on a button, the computer keyboard input option is gone (focus lost) and the user needs to click the "switch" again to get keyboard input. Sorry for being so elaborate about this, but if I manage this, the panel would be much more usable!

**Reply by Ted-juh** [reply]
I figured it was a z-order problem. I think grabKeyBoardFocus is key here. And what if you attach a script that at mouseUp the focus goes back to the Keyboard? The keyboard should always be in focus except for when you do a mouseclick then.

After some research: I see you are out of luck if you use normal buttons. There's no componentLuaMouseUp method to attach a script to. This means you have to use a custom component where a mouseUp method "is" supported. So, at mouseDown, grab focus on the component you want and do what you want to do with the mouse-clicks. At mouseUp give the focus back to the keyboard for that component and start typing again.

On the other hand, I think there's still a problem with the focus being reestablished. It still loses focus in between. Ok, tough one.  There are a lot of topics about focus, z-order, focustraverser, keypresslistener on the Juce Forums. It might help. If you can find a way to implement it in Ctrlr, that is. Some might be supported by Ctrlr, some might not be. And even if they are supported (.def createFocusTraverser etc etc), it would take me some time to find out how to use them.

**Reply by bijlevel** [reply]
I thought I had it solved by converting all the buttons to uiCustomComponent, so there is always one that has the focus, but I forgot on the panel is also an uiImageSlider which the user can click , so then that slider gets the focus. So problem not solved!

The main issue is still: how to write a method that, when called, sets always the focus to a certain (which one doesn't matter really) uiCustomComponent?

**Reply by Ted-juh** [reply]
setMouseClickGrabsKeyboardFocus(bool shouldgrabfocus). Set it to false if you don't want the component to grab keyboardfocus on mouseDown. Probably needs to be set in the mouseDown method of the uiImageSlider (or any component). Probably you need to setWantsKeyboardFocus again to the numeric pad. Not sure about that.

**Reply by bijlevel** [reply]
Thanks for your input, much appreciated, but I'm unable to make it work :-(. I'm getting nothing but "nil" errors. Probable because of my poor programming skills ;-)! I don't understand why it is so difficult to set the focus to a CustomComponent from within another component, in this case an ImageSlider.

**Reply by Ted-juh** [reply]
Is it possible to share some code? Or email me at tedjuh at Yahoo dot com to speak in our native tongue. Jeweetwel, Nederlands enzo :)

**Reply by bijlevel** [reply]
Yes please! I will send you an e-mail.

**Reply by damiensellier** [reply]
I don't know which version you use, probably @bijlevel 's latest release but the older macOS don't support mousedown trigger methods for uiImageButtons, uiImageSliders are ok though. I wasted like 3 hours last week because of that bug.

**Reply by bijlevel** [reply]
I'm using v5.3.198 on Mac and this is what the editor shows when having a uiImageButton on the canvas:

<img width="465" alt="Screenshot 2021-06-06 at 11 20 19" src="https://user-images.githubusercontent.com/11859363/120919343-61faae00-c6b9-11eb-9f83-ab77f41a8629.png">

So to me that looks like you can detect mouse actions and work with that in your methods. BUT I maybe wrong!

**Reply by damiensellier** [reply]
I don't remember exactly what was wrong, I'll check tomorrow.

**Reply by Godlike-Productions** [reply]
We might be able to throw all those number buttons in an invisible group and use that as the parent component for the getFocus.

**Reply by bijlevel** [reply]
From a personal contact with @DnaldeT I got a CustomComponent simulating a Alpha Dial, which was the only component I couldn't change to a CustomComponent. This code works OK, so I thought replacing all the buttons with CustomComponents would do the trick. As a quick test, today I changed the demo panel from @RomanKubiak. I added a second CustomComponent by copying the original one. But though the same methods/callback's are associated with the copied component, only the original component reacts to key presses when I click on the component with the mouse. When clicking the copied component, nothing happens. (BTW I also made a second CustomComponent from scratch, but no difference.) This is a mystery to me, because I was thinking that clicking on a component gives it automatically the focus! So if this simple test doesn't even work, I doubt if my original goal will ever be reached.

@Godlike-Productions maybe your approach can work, but don't forget we always need two "behaviours": clicking on a button AND pressing the associated key should both perform the same action.

**Reply by Ted-juh** [reply]
Simple solution but maybe not so easy to do: Create one Customcomponent that holds the alpha dial "and" the buttons. No matter if you click the alpha dial or the numeric pad, the component is in focus, thus receiving keypresses for the whole component. 

From the sounding of it, it's very close to my arpeggiator panel in which I determine the Mousedown on x,y coordinates to control a slider, set different buttons, and draw rectangles on 16 different lines. All in one Customcomponent.

**Reply by bijlevel** [reply]
Thanks again for your input! Simple enough (I hope), but as you indicate, one should also keep track of the x, y coördinates of the mouse click, because then you know which button, or more in general, which component did receive the click. A lot of work I think, and again, both ways (key press and mouse click) should be possible interchangeable.

**Reply by bijlevel** [reply]
Today I did build a simple panel from scratch and copied the methods of the keypress demo panel and the various uiCustomComponents get the focus when clicked and do recognise which key is pressed on the computer keyboard. So now the behaviour is as expected and I can start working on changing all the buttons on the panel to uiCustomComponent. An uiCustomComponent does also recognise a mouseUp event, so it can still behave as a regular button. This means I can preserve the original methods which deal with mouse clicks! A lot of work ahead, but doable, I hope.

**Reply by bijlevel** [reply]
As a result of my findings with the computer keyboard input, I post a small demo panel. A few remarks:

- **Technique**: ALL components are uiCustomComponents, because these are the only ones that recognise a keypress. One big uiCustomComponent covers (in this demo) the whole numeric keypad to increase the chance that a uiCustomComponent always has the focus. It turned out that one still can click on the components within this big one. Nice!;
- **Functionality**: as soon as the user clicks somewhere on the image of the keypad (or the dial) the key input functions;
- The midi messages are for the Kurzweil K2xxxR series instruments. I learned that CTRLR does automatically cast the values inserted in the midi message to the required Hex values. Nice!;
- I added the so called Alpha Dial, because the Kurzweil does have one. With the very kind help of @DnaldeT (and I believe he developed the code in collab with John Goodweather) I managed to use a uiCustomComponent for the dial. So no filmstrip needed! For the Kurzweil, the dial needed a 'spring' function, which was also possible with the code provided. Nice!
- **Issue one**: after loading the panel I couldn't find  way to activate a component (set the focus), so the whole thing only works after the user does click somewhere with the mouse. Knowing me, I always start clicking on everything, so that shouldn't be a big problem ;-). 
- **Issue two**: when exporting the panel as .bpanelz, NEVER do the components align with the background image. This is most irritating and is the subject of my next new discussion.

(https://github.com/RomanKubiak/ctrlr/files/6790575/K2x00R.key.input.demo_SWB_2021-07-09_11-50.bpanelz.zip)

**Comment by Ted-juh** [comment]
Had a good look at your panel. Dial works. Keyboard focus is there only when the component is clicked. Going over the Juce forums it seems you have to set all other components to "setWantsKeyboardFocus(false)"  except for the component that you want to. That one needs to be set to "true". I tried to test it but I think the "panel" itself grabs keyboard focus.  And I can't find a way to set it to false. (I might be completely off here.. but I think that is what is happening.)

However, I'm not entirely sold on the "extra" components for the buttons. One, it's an extra of 15 components. And two, because of your issue. So my advice is, keep it to one component. That way you won't run into issue 2. But how? I hear you ask.. Divide the custom component into regions, based on mouse coordinates, decide what should happen. I made a test panel to give you a headstart. 

The last thing to add, for now, the Numpad picture is attached to the panel background. Add it to the custom component (the big one) instead of the panel. (Same as you did with the alpha dial but without the rotation of course.)

[k2x00R test.zip](https://github.com/RomanKubiak/ctrlr/files/6795208/k2x00R.test.zip)

**Reply by bijlevel** [reply]
@DnaldeT Thanks so much for your input! I came across the Juce discusion about setting just one component 'true' for the keyboard focus and the rest to 'false', but that does not solve IMHO the fact that the focus is lost as soon as the user clicks elsewhere. But I admit I did not test this extensively.

I did look at your panel and of course with less components, less are misaligned, but still even if there is one, it is misaligned! Besides, I also think that in essence using components is coding wise not much different from your grid approach, because I can reuse the mouse down method for every button, so one method for all 15+ components. Also in my case the dial needs to be inside the image:

<img width="801" alt="Screenshot 2021-07-10 at 17 24 36" src="https://user-images.githubusercontent.com/11859363/125168042-c2db4180-e1a3-11eb-9084-8eb72e0dbd63.png">

and I'm not sure how to fit that into your grid approach.

Attaching the background image to a uiCustomComponent is a very interesting tip and I will certainly look into it! maybe it is the first step towards solving the annoying zoom issue. (Next time I will start a discussion about the zoom problem.)

**Reply by Ted-juh** [reply]
Maybe I don't get it but what exactly is misaligned? The panel I posted here still uses your original code. The numpad image on the background with one custom component placed over it. That those two do misalign is not a good thing, that's for sure. But if the numpad image is used as a background for the custom component itself (scalable) then there is no way it can be misaligned. Because the grid and picture scale with the component and thus the panel. As long as you don't use absolute coordinates, you should be fine.

The dial can be incorporated in the same custom component as the numpad. With some old-school math, the dial can be moved with the mouse only by clicking and dragging from the circular shape. 

I think I mentioned "offsets" in my panel. That way you can divide a custom component into different regions all with their own "grid" to define what happens on certain coordinates on mouseDown. Later tonight I'll make a testpanel to show you what I mean.

**Reply by Ted-juh** [reply]
Here is a quick mockup. Both the AlphaDial and the Numpad in one component. Here's in short what's going on behind the scenes:

The Component is divided into two regions, left and right. Left is the AlphaDial, right is the NumPad. The regions are detected on MouseMove (see label, upper right corner, which reads its state from a Memoryblock). If you are in region 0, the AlphaDial can be moved with MouseDrag. Its bounds are set to the whole left region. With some math, the bounds can be more restricted to the AlphaDial itself. If you move to region 1, then the grid for the Numpad is detecting which button you click (And MouseDrag won't move the AlphaDial.)

With some more work, the code for the keypresses can be fit in. Even the - + buttons under the AlphaDial can be done this way. And the best thing is.. you can resize the Mother Theresa out of it, the regions and grid for the NumPad stay in place. A short test with a restricted instance doesn't move anything. 

Mind that this is a really quick mockup. Lot of code to add to get the functionality you want. A lot to finetune, the rotation of the AlphaDial for instance. But the basics are there. If you really want to take it to another level, the whole right section of your panel can be done this way. (Up/down/left/right arrows, the AlphaDial, and the Numpad in one customComponent.)

[alphaDialNumpad.zip](https://github.com/RomanKubiak/ctrlr/files/6797363/alphaDialNumpad.zip)

**Reply by bijlevel** [reply]
@DnaldeT Thank you so much for your efforts and creativity to deal with what was initially a discussion about how CTRLR recognises computer keyboard input! But if I continue "your" way, I feel I have to re-code the whole panel and this is a project that involves 4 panels of which three are generally the same but the 4th has many functions that differ from the others. And because we were well on our way to (finally!) finish all the panels, re-coding them is ATM a bridge too far. Also this discussion went a bit off topic, becauseI think in general I did solve the issue of keypress input, only one issue remained: how to set the focus on a uiCustomComponent so the panel recognises keypress input from the moment it is loaded. I'm not sure if your approach does solve this. The issue of misalignment only comes up when I export the panel as .bpanelz and has to do with, I guess, careful tuning (or coding) the properties of uiPanelZoom, uiCanvasRectangle and uiPanelWidth/uiPanelHeight or even some other panel properties.
